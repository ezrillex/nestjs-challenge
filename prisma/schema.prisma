// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

// Looking for ways to speed up your queries, or scale easily with your serverless or edge functions?
// Try Prisma Accelerate: https://pris.ly/cli/accelerate-init

// npx prisma migrate dev --name the_name
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model Testing {
  id   String @id @default(uuid())
  test String
}

enum roles {
  customer
  manager
  admin
  anonymous // public
}

model Users {
  id                                 String     @id @default(uuid())
  first_name                         String     @db.VarChar(60)
  last_name                          String     @db.VarChar(120)
  email                              String     @unique
  password                           String     @db.Char(60) // hashed and salted password
  created_at                         DateTime   @default(now())
  role                               roles      @default(customer)
  failed_login_attempts              Int        @default(0)
  failed_login_attempts_timestamps   DateTime[] @default([])
  last_login_at                      DateTime?
  password_reset_requests            Int        @default(0)
  password_reset_requests_timestamps DateTime[] @default([])
  password_reset_token               String?
  password_last_updated              DateTime   @default(now())
}
