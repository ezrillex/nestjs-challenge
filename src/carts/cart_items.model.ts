import { Field, ID, Int, ObjectType } from '@nestjs/graphql';
import { Users } from '../auth/users.model';
import { ProductVariations } from '../products/product_variation/product-variations';

@ObjectType()
export class CartItems {
  @Field(() => ID)
  id: string;

  @Field(() => Users)
  cart_owner: Users;

  @Field(() => ProductVariations)
  product_variation: ProductVariations;

  @Field(() => Int)
  quantity: number;
}
