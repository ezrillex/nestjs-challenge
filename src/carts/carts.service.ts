import { BadRequestException, Injectable } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';

@Injectable()
export class CartsService {
  constructor(private readonly prisma: PrismaService) {}
  async AddToCart(variation_id: string, user_id: string, quantity: number) {
    const variation = await this.prisma.productVariations.count({
      where: { id: variation_id },
    });
    if (variation === 0) {
      throw new BadRequestException('Product Variation not found!');
    }

    const record = await this.prisma.cartItems.findFirst({
      where: {
        product_variation_id: variation_id,
        user_id: user_id,
      },
    });
    if (record) {
      return this.prisma.cartItems.update({
        where: { id: record.id },
        data: {
          quantity: quantity,
        },
      });
    } else {
      return this.prisma.cartItems.create({
        data: {
          cart_owner: {
            connect: { id: user_id },
          },
          product_variation: {
            connect: { id: variation_id },
          },
          quantity: quantity,
        },
      });
    }
  }

  async RemoveCartItem(cart_id: string) {
    const record = await this.prisma.cartItems.count({
      where: { id: cart_id },
    });
    if (record === 0) {
      throw new BadRequestException('Cart Item not found!');
    }

    const result = await this.prisma.cartItems.delete({
      where: { id: cart_id },
    });

    if (result) {
      return 'Cart Item deleted successfully.';
    } else {
      throw new BadRequestException('Unexpected error when deleting record!');
    }
  }

  async GetCartItems(user_id: string) {
    // TODO A resolve field can take care of a lot of this include stuff
    return this.prisma.cartItems.findMany({
      include: {
        cart_owner: {
          include: {
            CartItems: {
              include: {
                product_variation: true,
              },
            },
            likes_products: {
              include: {
                likes_product_variation: true,
              },
            },
          },
        },
        product_variation: {
          include: { images: true },
        },
      },
      where: {
        user_id: user_id,
        // hide deleted or private items.
        product_variation: {
          product: { is_deleted: false, is_published: true },
        },
      },
    });
  }
}
