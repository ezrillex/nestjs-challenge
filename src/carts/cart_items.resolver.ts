import {
  Args,
  Context,
  ID,
  Int,
  Mutation,
  Parent,
  Query,
  ResolveField,
  Resolver,
} from '@nestjs/graphql';
import { RequiresRole } from '../common/decorators/requires-role.decorator';
import { roles } from '@prisma/client';
import { ParseIntPipe, ParseUUIDPipe } from '@nestjs/common';
import { CartItems } from './cart_items.model';
import { CartsService } from './carts.service';
import { ProductsService } from '../products/products.service';
import { UsersService } from '../users/users.service';
import { ProductVariations } from '../products/product_variation/product-variations.model';
import { Users } from '../users/users.model';

@Resolver(() => CartItems)
export class CartItemsResolver {
  constructor(
    private readonly cartsService: CartsService,
    private readonly productsService: ProductsService,
    private readonly usersService: UsersService,
  ) {}

  // create or update
  @RequiresRole(roles.customer)
  @Mutation(() => CartItems, { nullable: true })
  async addToCart(
    @Args('variation_id', { type: () => ID }, ParseUUIDPipe)
    variation_id: string,
    @Args('quantity', { type: () => Int }, ParseIntPipe) quantity: number,
    @Context('req') request: Request,
  ): Promise<CartItems> {
    return await this.cartsService.addToCart(
      variation_id,
      request['user'].id,
      quantity,
    );
  }

  @RequiresRole(roles.customer)
  @Mutation(() => String, { nullable: true })
  async removeFromCart(
    @Args('product_variation_id', { type: () => String }, ParseUUIDPipe)
    product_variation_id: string,
    @Context('req') request: Request,
  ): Promise<string> {
    return this.cartsService.removeFromCart(
      product_variation_id,
      request['user'].id,
    );
  }

  @RequiresRole(roles.customer)
  @Query(() => [CartItems], { nullable: true })
  async getCartItems(@Context('req') request: Request): Promise<CartItems[]> {
    return this.cartsService.getCartItems(request['user'].id);
  }

  @ResolveField()
  async cart_owner(@Parent() cart_items: CartItems): Promise<Users> {
    const { id } = cart_items;
    return this.usersService.getUserByCartItem(id);
  }

  @ResolveField()
  async product_variation(
    @Parent() cart_items: CartItems,
  ): Promise<ProductVariations> {
    const { id } = cart_items;
    return this.productsService.ResolveProductVariationOnCartItems(id);
  }
}
