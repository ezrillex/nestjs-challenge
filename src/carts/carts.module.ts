import { Module } from '@nestjs/common';
import { CartsService } from './carts.service';
import { CartItemsResolver } from './cart_items.resolver';
import { PrismaService } from '../prisma/prisma.service';
import { ProductsService } from '../products/products.service';
import { UsersService } from '../users/users.service';

@Module({
  providers: [
    CartsService,
    CartItemsResolver,
    PrismaService,
    ProductsService,
    UsersService,
  ],
})
export class CartsModule {}
