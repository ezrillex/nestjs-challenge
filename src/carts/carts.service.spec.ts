import { Test, TestingModule } from '@nestjs/testing';
import { PrismaService } from '../prisma/prisma.service';
import { CartsService } from './carts.service';
import { CartItems } from '@prisma/client';

describe('CartsService', () => {
  let service: CartsService;
  let prismaService: PrismaService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [PrismaService, CartsService],
    }).compile();

    service = module.get<CartsService>(CartsService);
    prismaService = module.get<PrismaService>(PrismaService);
  });

  it('CartsService should be defined', () => {
    expect(service).toBeDefined();
  });

  it('PrismaService should be defined', () => {
    expect(prismaService).toBeDefined();
  });

  it('addToCart should be defined', () => {
    expect(service.addToCart).toBeDefined();
  });

  it('removeFromCart should be defined', () => {
    expect(service.removeFromCart).toBeDefined();
  });

  it('getCartItems should be defined', () => {
    expect(service.getCartItems).toBeDefined();
  });

  describe('addToCart', () => {
    it('throws an error when the product variation ID does not exist', async () => {
      await expect(
        service.addToCart(
          '2730fc05-6f87-49e5-8a41-559208048ebe',
          '2730fc05-6f87-49e5-8a41-559208048ebe',
          50,
        ),
      ).rejects.toThrowErrorMatchingSnapshot('product variation 404');
    });

    it('updates an existing cart item when a matching entry is found', async () => {
      jest.spyOn(prismaService.productVariations, 'count').mockResolvedValue(1);
      jest
        .spyOn(prismaService.cartItems, 'findUnique')
        .mockResolvedValue({ id: 'testing' } as CartItems);

      let updated = false;
      jest.spyOn(prismaService.cartItems, 'update').mockImplementation(() => {
        updated = true;
        return null;
      });

      await service.addToCart(
        '2730fc05-6f87-49e5-8a41-559208048ebe',
        '2730fc05-6f87-49e5-8a41-559208048ebe',
        3,
      );
      expect(updated).toEqual(true);
    });

    it('creates a new cart item when no existing entry is found for the provided data', async () => {
      jest.spyOn(prismaService.productVariations, 'count').mockResolvedValue(1);
      jest.spyOn(prismaService.cartItems, 'findFirst').mockResolvedValue(null);

      let created = false;
      jest.spyOn(prismaService.cartItems, 'create').mockImplementation(() => {
        created = true;
        return null;
      });

      await service.addToCart(
        '2730fc05-6f87-49e5-8a41-559208048ebe',
        '2730fc05-6f87-49e5-8a41-559208048ebe',
        3,
      );
      expect(created).toEqual(true);
    });
  });

  describe('removeFromCart', () => {
    it('throws an error when attempting to remove a cart item that does not exist', async () => {
      await expect(
        service.removeFromCart(
          '2730fc05-6f87-49e5-8a41-559208048ebe',
          '2730fc05-6f87-49e5-8a41-559208048ebe',
        ),
      ).rejects.toThrowErrorMatchingSnapshot(
        'tried to remove non existing cart items id',
      );
    });

    it('throws an error if the database delete operation fails without returning a deleted object', async () => {
      jest
        .spyOn(prismaService.cartItems, 'delete')
        .mockRejectedValue(new Error('testing'));
      jest.spyOn(prismaService.cartItems, 'count').mockResolvedValue(1);

      await expect(
        service.removeFromCart(
          '2730fc05-6f87-49e5-8a41-559208048ebe',
          '2730fc05-6f87-49e5-8a41-559208048ebe',
        ),
      ).rejects.toThrowErrorMatchingSnapshot(
        'delete called to db but deleted obj not returned, something is wrong',
      );
    });

    it('returns "Cart Item deleted successfully." upon successful deletion of the cart item', async () => {
      jest.spyOn(prismaService.cartItems, 'delete').mockResolvedValue({
        id: 'testing',
        user_id: 'testing',
        product_variation_id: 'testing',
        quantity: 3,
      });
      jest
        .spyOn(prismaService.cartItems, 'findUnique')
        .mockResolvedValue({ testing: 'test' } as any);

      await expect(
        service.removeFromCart(
          '2730fc05-6f87-49e5-8a41-559208048ebe',
          '2730fc05-6f87-49e5-8a41-559208048ebe',
        ),
      ).resolves.toEqual('Cart Item deleted successfully.');
    });
  });

  describe('getCartItems', () => {
    it('should call the database with the correct query for retrieving cart items', async () => {
      const spy = jest
        .spyOn(prismaService.cartItems, 'findMany')
        .mockResolvedValue(null);

      await service.getCartItems('2730fc05-6f87-49e5-8a41-559208048ebe');
      expect(spy.mock.calls).toMatchSnapshot('expected query get cart items');
    });
  });
});
