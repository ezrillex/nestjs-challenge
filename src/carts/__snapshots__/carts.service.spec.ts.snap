// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartsService removeFromCart throws an error if the database delete operation fails without returning a deleted object: Prisma delete error, not found 1`] = `
"
Invalid \`prisma.cartItems.delete()\` invocation:


An operation failed because it depends on one or more records that were required but not found. Record to delete does not exist."
`;
